src/spec-node/imageMetadata.ts

```diff
@@ -0,0 +1,375 @@
+import { extractDockerfile, findBaseImage, findUserStatement } from './dockerfileUtils';
+import { SubstituteConfig, SubstitutedConfig, DockerResolverParameters, inspectDockerImage, uriToWSLFsPath } from './utils';
+
+const pickConfigProperties: (keyof DevContainerConfig & keyof ImageMetadataEntry)[] = [
+	'onCreateCommand',
+	'updateContentCommand',
+	'postCreateCommand',
+	'postStartCommand',
+	'postAttachCommand',
+	'waitFor',
+	'customizations',
+	'mounts',
+	'containerEnv',
+	'containerUser',
+	'init',
+	'privileged',
+	'capAdd',
+	'securityOpt',
+	'remoteUser',
+	'userEnvProbe',
+	'remoteEnv',
+	'overrideCommand',
+	'portsAttributes',
+	'otherPortsAttributes',
+	'forwardPorts',
+	'shutdownAction',
+	'updateRemoteUserUID',
+	'hostRequirements',
+];
+
+const pickFeatureProperties: (keyof Feature & keyof ImageMetadataEntry)[] = [
+	'id',
+	'init',
+	'privileged',
+	'capAdd',
+	'securityOpt',
+	'entrypoint',
+	'mounts',
+	'customizations',
+];
+
+export interface ImageMetadataEntry {
+	id?: string;
+	init?: boolean;
+	privileged?: boolean;
+	capAdd?: string[];
+	securityOpt?: string[];
+	entrypoint?: string;
+	mounts?: (Mount | string)[];
+	customizations?: Record<string, any>;
+	onCreateCommand?: string | string[];
+	updateContentCommand?: string | string[];
+	postCreateCommand?: string | string[];
+	postStartCommand?: string | string[];
+	postAttachCommand?: string | string[];
+	waitFor?: DevContainerConfigCommand;
+	remoteUser?: string;
+	containerUser?: string;
+	userEnvProbe?: UserEnvProbe;
+	remoteEnv?: Record<string, string | null>;
+	containerEnv?: Record<string, string>;
+	overrideCommand?: boolean;
+	portsAttributes?: Record<string, PortAttributes>;
+	otherPortsAttributes?: PortAttributes;
+	forwardPorts?: (number | string)[];
+	shutdownAction?: 'none' | 'stopContainer' | 'stopCompose';
+	updateRemoteUserUID?: boolean;
+	hostRequirements?: HostRequirements;
+}
+
+export type MergedDevContainerConfig = MergedConfig<DevContainerFromImageConfig> | MergedConfig<DevContainerFromDockerfileConfig> | MergedConfig<DevContainerFromDockerComposeConfig>;
+
+type MergedConfig<T extends DevContainerConfig> = Omit<T, typeof replaceProperties[number]> & UpdatedConfigProperties;
+
+const replaceProperties = [
+	'customizations',
+	'entrypoint',
+	'onCreateCommand',
+	'updateContentCommand',
+	'postCreateCommand',
+	'postStartCommand',
+	'postAttachCommand',
+	'shutdownAction'
+] as const;
+
+interface UpdatedConfigProperties {
+	customizations?: Record<string, any[]>;
+	entrypoints?: string[];
+	onCreateCommands?: (string | string[])[];
+	updateContentCommands?: (string | string[])[];
+	postCreateCommands?: (string | string[])[];
+	postStartCommands?: (string | string[])[];
+	postAttachCommands?: (string | string[])[];
+	shutdownAction?: 'none' | 'stopContainer' | 'stopCompose';
+}
+
+export function mergeConfiguration(config: DevContainerConfig, imageMetadata: ImageMetadataEntry[]): MergedDevContainerConfig {
```

Comment:
What is the purpose or advantage of having a type `MergedDevContainerConfig` that's distinct form `DevContainerConfig`?